cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRunner.java

```diff
@@ -63,37 +70,54 @@
 
 import java.io.Closeable;
 import java.io.File;
-import java.util.ArrayList;
-import java.util.List;
+import java.io.IOException;
+import java.util.Deque;
+import java.util.LinkedList;
 import java.util.concurrent.TimeUnit;
 import javax.annotation.Nullable;
 
 /**
- * Runs {@link Spark} programs
+ * The {@link ProgramRunner} that executes Spark program.
  */
-public class SparkProgramRunner extends AbstractProgramRunnerWithPlugin {
+final class SparkProgramRunner extends AbstractProgramRunnerWithPlugin implements ProgramClassLoaderFilterProvider {
 
   private static final Logger LOG = LoggerFactory.getLogger(SparkProgramRunner.class);
+  static final FilterClassLoader.Filter SPARK_PROGRAM_CLASS_LOADER_FILTER = new FilterClassLoader.Filter() {
+
+    final FilterClassLoader.Filter defaultFilter = FilterClassLoader.defaultFilter();
+
+    @Override
+    public boolean acceptResource(String resource) {
+      return resource.startsWith("co/cask/cdap/api/spark/") || resource.startsWith("scala/")
+        || resource.startsWith("org/apache/spark/") || resource.startsWith("akka/")
+        || defaultFilter.acceptResource(resource);
+    }
+
+    @Override
+    public boolean acceptPackage(String packageName) {
```

Comment:
If `acceptPackage()` only delegates to `defaultFilter`, then for packages `co/cask/cdap/api/spark`, `scala`, etc. this method will return false, even though `acceptResource()` method would return true for classes in those packages.